### ISSUE TITLE
Golden testing for expected failures

### ISSUE THREAD

**Issue by david-christiansen** [gelisam/klister#42]
The `non-examples` directory contains files we expect to fail. We should augment the testing infrastructure to ensure they continue to fail in the way we expect.

**Comment by doyougnu** [comment]
started on this:

```
  which-problem:                                     OK (1.20s)
    incorrect-context:                                 FAIL (0.04s)
      tests/Golden.hs:53:
      expected:
      #[incorrect-context.kl:5.10-5.22]<(define x 5)>:
        Used in a position expecting an expression
        but is valid in a position expecting a top-level declaration or example
      ExitFailure 1
      
      got:
      #[incorrect-context.kl:5.10-5.22]<(define x 5)>:
        Used in a position expecting an expression
        but is valid in a position expecting a top-level declaration or example
    bound-vs-free:                                     OK (0.04s)
    meta-macro:                                        OK (0.14s)
    integer-syntax:                                    OK (0.05s)
    import:                                            OK (0.03s)
```

but then I noticed that `cabal test --test-option='--accept'` returns:
```
Running 1 test suites...
Test suite klister-tests: RUNNING...
Invalid option `--accept'

Usage: klister-tests [-p|--pattern PATTERN] [-t|--timeout DURATION] 
                     [-l|--list-tests] [-j|--num-threads NUMBER] [-q|--quiet] 
                     [--hide-successes] [--color never|always|auto] 
                     [--ansi-tricks ARG] [--hedgehog-replay ARG] 
                     [--hedgehog-show-replay ARG] [--hedgehog-tests ARG] 
                     [--hedgehog-discards ARG] [--hedgehog-shrinks ARG] 
                     [--hedgehog-retries ARG]

Test suite klister-tests: FAIL
```

which means that `tasty-golden` isn't being used at all. Is there any reason? Are is it time to actually use it?

**Comment by doyougnu** [comment]
it looks pretty good to me:
```
    custom-module-test:                                OK (0.78s)
    which-problem:                                     OK (1.02s)
    incorrect-context:                                 FAIL (0.04s)
      Test output was different from 'examples/failing-examples/incorrect-context.golden'. It was:
      #[incorrect-context.kl:5.10-5.22]<(define x 5)>:
        Used in a position expecting an expression
        but is valid in a position expecting a top-level declaration or example
    bound-vs-free:                                     OK (0.04s)
    meta-macro:                                        OK (0.16s)
    integer-syntax:                                    OK (0.05s)
    import:                                            OK (0.05s)

1 out of 124 tests failed (48.21s)

Test suite klister-tests: FAIL
```

where the golden file was:
```
#[incorrect-context.kl:5.10-5.22]<(define x 5)>:
  Used in a position expecting an expression
  but is valid in a position expecting a top-level declaration or example
ExitFailure 1
```
note the extra `ExitFailure 1` because I copied it from the command line

**Comment by gelisam** [comment]
> we never use a function that creates a `TestTree` from `Tasty.Golden` only a `TestTree` from `HUnit`. Is there any reason? Are is it time to actually use it?

My [original version](https://github.com/gelisam/klister/commit/24982804bcd80a0283ff8a54d9be38a1e11f834a#diff-79a19ed5a00673aa6f567829ae6f5622153a8ef7c523a115738c4a05083688b8) of the golden tests used `Test.Tasty.Golden.goldenVsString`, because it was the easiest way to implement golden tests, but the error messages were not very useful:

```
All tests
  Golden tests
    lang:                                              FAIL (0.02s)
      Test output was different from 'examples/lang.golden'. It was:
      #[lang.kl:5.30-5.31]<ø> : Syntax
```

I fixed that problem twice:

1.  [Once](https://github.com/gelisam/klister/commit/682a220f13babbd893f51f3004203b230fc13699) by switching to `Test.Tasty.Golden.goldenVsStringDiff`, producing the following output:
    
    ```
    All tests
      Golden tests
        lang:                                              FAIL (0.02s)
          Test output was different from 'examples/lang.golden'. Output of ["diff","-u","examples/lang.golden","/tmp/lang2846965-21.actual"]:
          --- examples/lang.golden	2023-03-02 21:31:34.059432901 -0500
          +++ /tmp/lang2846965-21.actual	2023-03-02 21:53:27.011355802 -0500
          @@ -1 +1 @@
          -#[lang.kl:6.30-5.31]<ø> : Syntax
          +#[lang.kl:5.30-5.31]<ø> : Syntax
    ```
    
2. [A second time](https://github.com/gelisam/klister/commit/73562e857a5f699d4ee15bb36d5024a82f934fe2) by reimplementing the (very simple) logic of a golden test, producing the following output:
    
    ```
    All tests
      Golden tests
        lang:                                              FAIL (0.04s)
          tests/Golden.hs:52:
          expected:
          #[lang.kl:6.30-5.31]<ø> : Syntax
      
          got:
          #[lang.kl:5.30-5.31]<ø> : Syntax
    ```

Why two implementations? It just so happens that implementation number 1 was attached to a branch which never got merged, and I had forgotten about implementation number 1 when I worked on implementation number 2.

Both implementations seem fine to me; but I suspect you like implementation number 1 better because it happens to make use of `Test.Tasty.Golden` and thus probably supports `--accept`. I didn't know about `--accept`, I usually run `./examples/regenerate-golden-files.sh` (or rather a locally modified version which uses cabal instead of stack) to delete all the golden files and then regenerate them using `cabal test`.

> it looks pretty good to me, also it simplifies the testsuite:

It doesn't look that good to me because the error message only shows the actual value, not the expected value. But I guess that doesn't really matter, because my workflow is to run `./examples/regenerate-golden-files.sh` and then use `git diff` to see if the changes match what I expect. So if you plan to look at the error message more closely than I do, it would make more sense for you to pick an error message format which suits your tasts, not mine :)

**Comment by doyougnu** [comment]
> but the error messages were not very useful:

I agree. 

I think I'll whip up something using `tasty-golden`. I'm hesitant to keep code that reimplements a commonly used library unless that library has some significant issue. Only because then we don't benefit from any future updates or features and there is no need to reinvent the wheel as long as the lib can do what we need.

**Comment by gelisam** [comment]
Fixed by #210
